libconcurrent/primitives/threadtools.c

```diff
@@ -59,33 +75,79 @@ inline static void *kthreadWrapper(void *arg) {
     return NULL;
 }
 
-inline uint32_t synchPreferedCoreOfThread(uint32_t pid) {
-    uint32_t prefered_core = 0;
+inline uint32_t synchPreferredCoreOfThread(uint32_t pid) {
+    uint32_t preferred_core = 0;
+
+    if (__schedule_policy == SYNCH_THREAD_PLACEMENT_FLAT) {
+        preferred_core = pid;
+    } else {
 #ifdef SYNCH_NUMA_SUPPORT
-    int ncpus = numa_num_configured_cpus();
-    int nodes = numa_num_task_nodes();
-    int node_size = ncpus / nodes;
-
-    if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) {
-        int half_node_size = node_size / 2;
-        int offset = 0;
-        uint32_t half_cpu_id = pid;
-
-        if (pid >= ncpus / 2) {
-            half_cpu_id = pid - ncpus / 2;
-            offset = ncpus / 2;
+        uint32_t ncpus = numa_num_configured_cpus();
+        uint32_t nodes = numa_num_task_nodes();
+        uint32_t node_size = ncpus / nodes;
+
+        if (__schedule_policy == SYNCH_THREAD_PLACEMENT_NUMA_SPARSE) {
+            if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) { // SMT or HyperThreading detected
+                uint32_t half_node_size = node_size / 2;
+                uint32_t offset = 0;
+                uint32_t half_cpu_id = pid;
+
+                if (pid >= ncpus / 2) {
+                    half_cpu_id = pid - ncpus / 2;
+                    offset = ncpus / 2;
+                }
+                preferred_core = (half_cpu_id % nodes) * half_node_size + half_cpu_id / nodes;
+                preferred_core += offset;
+            } else preferred_core = ((pid % nodes) * node_size);
```

Comment:
This will possibly oversubscribe cpus, right?
If this is the intention you can add it to the documentation

Reply:
Nop. It actually, tries to find if the processor supports SMT or not.
In case that it supports SMT, linux sets the SMT threads in the upper range of cores (e.g., if the total is 128 cores, cores with numbers 64-127 are the SMT threads).
Linux actually implements a weird numbering on numbering processing cores.

Reply:
I mean here: `preferred_core = ((pid % nodes) * node_size);`
Lets say:
1) num of pids = 16
2) nodes = 4
3) node_size = 4

then `(pid % nodes)` can be either 0, 1, 2 or 3
and `preferred_core` can be either 0, 4, 8, or 12.
Which means that we oversubscribe certain cores.

this is useful if the number of pids is much less than the number of total cores, and if we want to fully utilize the memory bandwidth, right? If yes, then you can specify that in the doc,

Reply:
Let me check this. It's the case where the machine is NUMA but without SMT support. We may have a bug here. This is a rare case, since NUMA without SMT are not very common machines and the CI tools cannot touch this part of code. I will come back on this.